### PR TITLE
pg_waldump: Added option --last-valid-walname

### PR DESCRIPTION
**Pull request by Annu149**
**Context:**
As suggested [here](https://github.com/greenplum-db/gpdb/pull/17126#issuecomment-2014060576) we need to fetch the last valid wal segment to calculate the remaining bytes of startup recovery. So added option --last-valid-walname to pg_waldump to emit the last valid wal segment for the timeline provided with -t.  

**New option:**
` --last-valid-walname     output last valid wal segment for given timeline`

**Usage:**
`pg_waldump -p <pg_wal_path> -t <timeline> --last-valid-walname`

**Example:**
```
pg_waldump -p ./gpAux/gpdemo/datadirs/dbfast3/demoDataDir2/pg_wal/ -t 15 --last-valid-walname 
000000010000000000000003
```
Testing: Added TAP tests for pg_waldump

Code changes borrowed from: https://www.postgresql.org/message-id/attachment/129055/v5-0001-Introduce-feature-to-start-WAL-receiver-eagerly.patch

Suggested-by : Soumyadeep Chakraborty <soumyadeep2007@gmail.com>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by soumyadeep2007** [comment]
Responding to offline questions here, so that others can follow along, and we have a record of the decisions taken:

> pg_waldump fails if no path is provided with -p or invalid timeline is provided. Should we do the same in case no valid wal file is found for given timeline?

> What if no wal segment is found for given timeline?
Should we print a message that 'No valid wal file found' with non-zero exit code?

Yes, just call `fatal_error` for this case, with error message: `could not find valid WAL segment for timeline \"%d\"`. `fatal_error` implicitly returns exit(EXIT_FAILURE).

> Output format?
> the output looks like below-
> ```pg_waldump -p ./gpAux/gpdemo/datadirs/dbfast3/demoDataDir2/pg_wal/ -t 15 --last-valid-wal 
> 3```

Have, `LastValidWalSegment` return `XLogSegno`, but print the full xlog filename utilizing `XLogFilename` in the callsite.

> flag combinations?
> Right now the options -p and -t are needed with --last-valid-wal. What should we do if  --last-valid-wal used with other options?

Since we just print the name of the log file, we aren't really dumping the contents (unlike the other pg_waldump options), the other flags don't really apply here. We can safely ignore the other flags. We can just add a validation that the timeline id is explicitly specified (and if not just `pg_log_error("...") and goto bad_argument;`). `-p` need not be specified btw and can be populated from defaults.

To make things clearer, `--last-valid-wal` -> `--last-valid-walname`

> Shorthand for --last-valid-wal -> -w ? Please let me know if any other suggestions

Lets not provide the short option (avoids future conflict with upstream, in case upstream introduces a -w). So, lets just have `--last-valid-wal`.

File: src/bin/pg_waldump/pg_waldump.c [hunk 1]
```diff
@@ -1015,6 +1128,12 @@ main(int argc, char **argv)
 		}
 	}
 
+	if(last_valid_wal)
+	{
+		fprintf(stdout,"%lu\n", LastValidWalSegment(private.timeline, private.inpath));
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:1133]: just printf should suffice.

File: src/bin/pg_waldump/pg_waldump.c [hunk 2]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+                 */
+                continue;
+            }
+            startsegno = (startsegno == -1) ? logSegNo : Min(startsegno, logSegNo);
+            endsegno = (endsegno == -1) ? logSegNo : Max(endsegno, logSegNo);
+        }
+    }
+    closedir(dir);
+	
+	if(startsegno == -1 && endsegno == -1){
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:870]: Please follow the PG code style: Space between `if and (` and `{` on next line, `/* */` for single line comments.

File: src/bin/pg_waldump/pg_waldump.c [hunk 3]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+        private.timeline = currentTLI;
+        private.startptr = startptr;
+        private.endptr = InvalidXLogRecPtr;
+        private.endptr_reached = false;
+		private.inpath = directory;
+
+        state = XLogReaderAllocate( wal_segment_size, XLogDumpReadPage, &private);
+
+        if (!state)
+			fatal_error("Out of memory, failed while allocating a WAL reading processor.");
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:897]: "Out of memory" should suffice.

File: src/bin/pg_waldump/pg_waldump.c [hunk 4]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+    XLogSegNo	startsegno = -1;
+    XLogSegNo	endsegno = -1;
+
+    /* Find the latest and earliest WAL segments in pg_wal */
+	dir = opendir(directory);
+	if(dir==NULL){
+		pg_log_error("invalid path provided");
+		fprintf(stderr, _("Try \"%s --help\" for more information.\n"), progname);
+		exit(1);
+	}
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:847]: Use `fatal_error` here.

File: src/bin/pg_waldump/pg_waldump.c [hunk 5]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+            startsegno = (startsegno == -1) ? logSegNo : Min(startsegno, logSegNo);
+            endsegno = (endsegno == -1) ? logSegNo : Max(endsegno, logSegNo);
+        }
+    }
+    closedir(dir);
+	
+	if(startsegno == -1 && endsegno == -1){
+		// No valid wal segment found for given timeline.
+		fprintf(stdout, "No valid wal segment found for given timeline\n");
+		return 0;
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:873]: Just use fatal_error

File: src/bin/pg_waldump/pg_waldump.c [hunk 6]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+    while (endsegno >= startsegno)
+    {
+        XLogReaderState *state;
+        XLogRecPtr	startptr;
+        char		xlogfname[MAXFNAMELEN];
+		
+		XLogSegNoOffsetToRecPtr(endsegno, 0, wal_segment_size, startptr);
+		XLogFileName(xlogfname, currentTLI, endsegno, wal_segment_size);
+
+        XLogDumpPrivate private;
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:886]: Nit: Code style: Declare local variable at top of scope (together with the other variables)

File: src/bin/pg_waldump/pg_waldump.c [hunk 7]
```diff
@@ -819,6 +825,105 @@ usage(void)
...
+        if (!state)
+			fatal_error("Out of memory, failed while allocating a WAL reading processor.");
+
+        /*
+         * Read the first page of the current WAL segment and validate it by
+         * inspecting the page header.
+         */
+        Assert(!XLogRecPtrIsInvalid(startptr));
+
+        /* Begin at the passed-in record pointer. */
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:905]: This wasn't passed in.

File: src/bin/pg_waldump/t/001_basic.pl [hunk 8]
```diff
@@ -2,9 +2,36 @@
...
 use PostgreSQL::Test::Utils;
 use Test::More;
+use PostgreSQL::Test::Cluster;
 
 program_help_ok('pg_waldump');
 program_version_ok('pg_waldump');
 program_options_handling_ok('pg_waldump');
 
+my $node = PostgreSQL::Test::Cluster->new('main');
+
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/t/001_basic.pl:12]: We should have the following test cases:

1. No walfile for given timeline exists, but walfiles from other timelines do.
2. Use pg_switch_wal to generate multiple WAL files, and ensure that the latest one is returned.
3. Same as 2, but add some invalid WAL files with higher and valid WAL file name but having size = 0 (this is exactly why we need to validate the page header etc).
4. Same as 2, but add some invalid WAL files with higher and valid WAL file name but having corrupt header (this is exactly why we need to validate the page header etc).
5. Add some other non-walfiles to ensure that those aren't considered (basically validating `IsXLogFileName`)

File: src/bin/pg_waldump/t/001_basic.pl [hunk 9]
```diff
@@ -2,9 +2,110 @@
...
+# Simulate corrupt header
+my $corrupt_header = "\x00" x $header_size; # Filling the header with null bytes
+print $fh $corrupt_header;
+close $fh;
+
+$node->command_checks_all(['pg_waldump', '-p', "$pgdata/pg_wal/", '-t', '1','--last-valid-walname'],
+    1,
+    [qr{^$}],
+    [qr/pg_waldump: fatal: could not read from log file 000000010000000000000004, offset 24: read 0 of 32744/],
+	'pg_waldump fails if some files can not be read');
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/t/001_basic.pl:91]: This test currently covers: `XLogDumpXLogRead` erroring out during `read`, as its trying to read 32K but can't as the wal file contains only 24 bytes correct? That is probably not the intent of this test.

This is a good test to have to see if pg_waldump is erroring out due to a read failure. But we should rename this to reflect that aspect. Also, as we had discussed before, we should also have coverage for the case where:
```
        /* Validate the PageHeader, if valid then our search for last valid wal file ends here. */
        if (XLogReaderValidatePageHeader(state, startptr, state->readBuf))
        {
            XLogReaderFree(state);
            return endsegno;
        }

        XLogReaderFree(state); <------------ execution comes here and we loop to the next
        endsegno--;
```

One of the ways to test this path is probably to have at least 32k bytes in the WAL file, so that `read()` doesn't fail. And then you can simulate a bad header by modifying the magic byte, so that `if (hdr->xlp_magic != XLOG_PAGE_MAGIC)` is tripped.

File: src/bin/pg_waldump/pg_waldump.c [hunk 10]
```diff
@@ -888,13 +983,19 @@ main(int argc, char **argv)
...
 	if (argc <= 1)
 	{
 		pg_log_error("no arguments specified");
 		goto bad_argument;
 	}
 
-	while ((option = getopt_long(argc, argv, "be:fn:p:r:s:t:x:z",
+	
+
+	while ((option = getopt_long(argc, argv, "be:fn:p:r:s:wt:x:z",
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:998]: `wt -> w:t`

File: src/bin/pg_waldump/pg_waldump.c [hunk 11]
```diff
@@ -967,6 +1069,10 @@ main(int argc, char **argv)
 					pg_log_error("could not parse timeline \"%s\"", optarg);
 					goto bad_argument;
 				}
+				timeline_specified = true;
+				break;
+			case 131:
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:1074]: Since we are specifying `-w` in getoptlong, why not specify it here too instead of using `131`?

Comment (Annu149) [reply src/bin/pg_waldump/pg_waldump.c:1074]: Removed -w as we don't want any short option for --last-valid-walname

File: src/bin/pg_waldump/t/001_basic.pl [hunk 12]
```diff
@@ -2,9 +2,110 @@
...
+my $current_walfile = $node->safe_psql('postgres', "SELECT pg_walfile_name('$current_wal_lsn');");
+
+my ($last_valid_walname, $stderr) = run_command(
+	[
+		'pg_waldump', '-p', "$pgdata/pg_wal/", '-t', '1', '--last-valid-walname'
+	]);
+ok ($last_valid_walname eq $current_walfile);
+
+# Create a wal file with zero size
+my $wal_file_name = "$pgdata/pg_wal/000000010000000000000002";
```
[the remark this answers is not in the document]
Comment (Annu149) [reply src/bin/pg_waldump/t/001_basic.pl]: Updated

File: src/bin/pg_waldump/pg_waldump.c [hunk 13]
```diff
@@ -888,12 +983,18 @@ main(int argc, char **argv)
...
+	bool timeline_specified = false;
+
 	if (argc <= 1)
 	{
 		pg_log_error("no arguments specified");
 		goto bad_argument;
 	}
 
+	
+
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:997]: Unnecessary whitespace diff

File: src/bin/pg_waldump/pg_waldump.c [hunk 14]
```diff
@@ -842,6 +936,7 @@ main(int argc, char **argv)
 		{"rmgr", required_argument, NULL, 'r'},
 		{"start", required_argument, NULL, 's'},
 		{"timeline", required_argument, NULL, 't'},
+		{"last-valid-walname", no_argument, NULL,131},
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/pg_waldump.c:939]: NULL,131 -> NULL, 131

File: src/bin/pg_waldump/pg_waldump.c (hunk 11 above)
Comment (soumyadeep2007) [reply src/bin/pg_waldump/pg_waldump.c:1074]: Got it!

File: src/bin/pg_waldump/t/001_basic.pl [hunk 15]
```diff
@@ -2,9 +2,126 @@
...
+my $current_walfile = $node->safe_psql('postgres', "SELECT pg_walfile_name('$current_wal_lsn');");
+
+my ($last_valid_walname, $stderr) = run_command(
+	[
+		'pg_waldump', '-p', "$pgdata/pg_wal/", '-t', '1', '--last-valid-walname'
+	]);
+ok ($last_valid_walname eq $current_walfile, 'last valid walfile is the most current walfile');
+
+# Create a wal file with zero size
+my $wal_file_name = "$pgdata/pg_wal/000000010000000000000004";
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/t/001_basic.pl:67]: Isn't the current walfile 04 (since you switched 3 times)? Then shouldn't this new walfile be 05?
Can you assert the contents of the wal directory here?

File: src/bin/pg_waldump/t/001_basic.pl [hunk 16]
```diff
@@ -2,9 +2,126 @@
...
+$node->command_checks_all(['pg_waldump', '-p', "$pgdata/pg_wal/", '-t', '1','--last-valid-walname'],
+    1,
+    [qr{^$}],
+    [qr/pg_waldump: fatal: could not read from log file 000000010000000000000004, offset 0: read 0 of 32768/],
+	'pg_waldump fails if any file is of size 0');
+
+rmtree($wal_file_name);
+
+# Create a walfile having corrupt header having size of 24 bytes
+my $corrupt_wal_file = "$pgdata/pg_wal/000000010000000000000005";
```
Comment (soumyadeep2007) [inline src/bin/pg_waldump/t/001_basic.pl:80]: Since you are removing 04 above, why not just recreate the file again, instead of adding 05, as that makes it confusing.

Just have 1 variable: `corrupt_walfile` and use it for this scenario and the reuse it for the one below.

File: doc/src/sgml/ref/pg_waldump.sgml [hunk 17]
```diff
@@ -160,6 +160,16 @@ PostgreSQL documentation
       </listitem>
      </varlistentry>
 
+     <varlistentry>
+      <term><option>-w <replaceable>timeline</replaceable></option></term>
```
Comment (soumyadeep2007) [inline doc/src/sgml/ref/pg_waldump.sgml:164]: Remove the -w from here.

**Review by soumyadeep2007 — APPROVED** [review]
LGTM! Mostly minor outstanding comments.

File: doc/src/sgml/ref/pg_waldump.sgml [hunk 18]
```diff
@@ -160,6 +160,15 @@ PostgreSQL documentation
       </listitem>
      </varlistentry>
 
+     <varlistentry>
+      <term><option>--last-valid-wal=<replaceable>timeline</replaceable></option></term>
+      <listitem>
+       <para>
+        Output last valid wal segment for given timeline.
```
Comment (shrakesh) [inline doc/src/sgml/ref/pg_waldump.sgml:167]: nit: Output->print / Display /show

File: src/bin/pg_waldump/pg_waldump.c [hunk 19]
```diff
@@ -819,6 +822,97 @@ usage(void)
...
+                 */
+                continue;
+            }
+            startsegno = (startsegno == -1) ? logSegNo : Min(startsegno, logSegNo);
+            endsegno = (endsegno == -1) ? logSegNo : Max(endsegno, logSegNo);
+        }
+    }
+    closedir(dir);
+	
+	if (startsegno == -1 && endsegno == -1)
```
Comment (shrakesh) [inline src/bin/pg_waldump/pg_waldump.c:865]: should it be || or && . although in an ideal scenario either both will be -1 or have some value. just wants to check whats the expectation here.

**Review by shrakesh — APPROVED** [review]
LGTM 
added a few minor comments.

File: doc/src/sgml/ref/pg_waldump.sgml (hunk 18 above)
Comment (Annu149) [reply doc/src/sgml/ref/pg_waldump.sgml:167]: Have followed existing format of help text
